In [150]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Lasso, Ridge
from sklearn.linear_model import LogisticRegression 
%matplotlib inline
%config InlineBakend.figure_format = 'svg'
from sklearn.svm import SVC
from sklearn.metrics import r2_score

## Задание №1

In [2]:
boston = load_boston()

In [3]:
data = boston.data

In [4]:
target = boston.target

In [5]:
feature_names = boston.feature_names

In [6]:
X = pd.DataFrame(data, columns=feature_names)

In [7]:
y = pd.DataFrame(target, columns=['price'])

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.3, random_state=42)

In [9]:
lr = LinearRegression()

In [10]:
lr.fit(X_train,y_train)

LinearRegression()

In [11]:
y_pred = lr.predict(X_test)
y_pred.shape

(152, 1)

In [12]:
check_test = pd.DataFrame({"y_test": y_test["price"],
                           "y_pred": y_pred.flatten(),})

In [13]:
mean_squared_error(check_test["y_pred"], check_test["y_test"])

21.517444231177368

In [14]:
r2_score(y_test,y_pred)

0.7112260057484912

## Задание №2

In [15]:
model = RandomForestRegressor(n_estimators=1000, random_state=42, max_depth=12)

In [16]:
model.fit(X_train, y_train.values[:,0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [17]:
y_pred_model = model.predict(X_test)

In [26]:
check_test = pd.DataFrame({"y_test": y_test["price"],
                           "y_pred_model": y_pred_model.flatten(),})
check_test

y_test  y_pred_model
173    23.6     22.806412
274    32.4     31.131464
491    13.6     16.339125
72     22.8     23.810726
452    16.1     17.139521
..      ...           ...
441    17.1     13.521640
23     14.5     15.112621
225    50.0     42.808700
433    14.3     15.586103
447    12.6     16.072425

[152 rows x 2 columns]

In [19]:
r2_score(y_test,y_pred_model)

0.87472606157312

#### Random Forest дает меньше погрешности и лучше чем алгоритм линейной регрессии (0.875 > 0.711) 

### Задание №3

In [21]:
model.feature_importances_

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [ ]:
sum(model.feature_importances_)

In [ ]:
sorted(model.feature_importances_,reverse=True)[0:2]

In [101]:
model.feature_importances_

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [25]:
feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [104]:
dd =  pd.DataFrame({"importance": model.feature_importances_, "feature_names": feature_names}).sort_values("importance")

In [105]:
dd[["feature_names", "importance"]].tail(2)

feature_names  importance
5             RM    0.402682
12         LSTAT    0.415847

### Задание 4

In [107]:
data_cc = pd.read_csv("creditcard.csv")

In [109]:
df_cc = pd.DataFrame(data_cc)

In [113]:
df_cc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [117]:
df_cc.value_counts(normalize=True)

Time      V1          V2          V3         V4         V5         V6         V7          V8          V9         V10        V11        V12        V13        V14        V15        V16        V17        V18        V19        V20        V21         V22        V23        V24        V25        V26        V27        V28        Amount  Class
163152.0  -1.203617    1.574009    2.889277   3.381404   1.538663   3.698747   0.560211   -0.150911    0.124136   4.220998   1.384569  -0.706897  -0.256274  -1.562583   1.692915  -0.787338  -0.226776  -0.412354   0.234322   1.385597  -0.366727    0.522223  -0.357329  -0.870174  -0.134166   0.327019  -0.042648  -0.855262  1.51    0        0.000063
          -1.196037    1.585949    2.883976   3.378471   1.511706   3.717077   0.585362   -0.156001    0.122648   4.217934   1.385525  -0.709405  -0.256168  -1.564352   1.693218  -0.785210  -0.228008  -0.412833   0.234834   1.375790  -0.370294    0.524395  -0.355170  -0.869790  -0.133198   0.327804  -0.035702  -0

In [115]:
pd.options.display.max_columns = 100

In [116]:
df_cc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [123]:
X = df_cc.drop(["Class"], axis=1)

In [125]:
y = pd.Series(df_cc["Class"])

In [134]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.3, random_state=30, stratify=y)

In [135]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((3322, 30), (278209, 30), (3322,), (278209,))

In [136]:
parameters = [{'n_estimators': [10, 15],
'max_features': np.arange(3, 5),
'max_depth': np.arange(4, 7)}]

In [140]:
model = GridSearchCV(estimator=RandomForestClassifier(random_state=100), 
                     param_grid = parameters,
                     scoring= 'roc_auc',
                     cv=3)

In [141]:
model.fit(X_train, y_train)

GridSearchCV(cv=3, estimator=RandomForestClassifier(random_state=100),
             param_grid=[{'max_depth': array([4, 5, 6]),
                          'max_features': array([3, 4]),
                          'n_estimators': [10, 15]}],
             scoring='roc_auc')

In [154]:
model.best_params_

{'max_depth': 6, 'max_features': 4, 'n_estimators': 15}

In [148]:
y_pred_GSC = model.predict_proba(X_test)

In [149]:
y_pred_GSC

array([[0.99023842, 0.00976158],
       [0.98788017, 0.01211983],
       [0.99164828, 0.00835172],
       ...,
       [0.96468824, 0.03531176],
       [0.99149621, 0.00850379],
       [0.92571471, 0.07428529]])

In [152]:
y_pred_proba = y_pred_GSC[:,1]
y_pred_proba

array([0.00976158, 0.01211983, 0.00835172, ..., 0.03531176, 0.00850379,
       0.07428529])

In [153]:
roc_auc_score(y_test, y_pred_proba)

0.9856801503189965